lib/io/ConcurentPromiseManager.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * A generic class to manage typed operation (Promises) in a pooling manner.
+ * useful for executing concurrent operations with control of the size of the pool.
+ */
+export class ConcurentPromiseManager<T> {
+  // Number of asyncronous operations
+  public readonly poolSize: number;
+  // Queue of operations
+  private operationQueue: IOperation<T>[];
+  // Results of the operations, might not be in the same order as their insertion in the queue
+  public readonly results: IResult<T>[] = [];
+
+  public constructor(poolSize: number, initialOperations: IOperation<T>[]) {
```

Comment:
I would not add this `initialOperations` param, as this allows skipping the poolSize check.